### PR TITLE
Image kinds re-design

### LINKED ISSUE #1879: Image view type
In Vulkan, creating an image view requires a kind of view to be specified. I.e. A 2D array texture can be seen with a 2D view. D3D12 has similar notation. We are currently deriving the view type from the image kind, which has a problem of potentially mismatching shader expectations. For instance, we are not able to use different-kind images to provide views for the same descriptor.

Solution: introduce a `ViewKind`/`ViewType` enum (matching Vulkan's 1:1 preferably) that is passed as an extra parameter to `create_image_view`.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #1879 

~~Early WIP opened for feedback on the new HAL image API bits.~~ Ready for review now.

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: Vulkan

### CODE REVIEW

File: src/backend/vulkan/src/lib.rs [hunk 1]
```diff
@@ -143,7 +143,7 @@ impl Instance {
             application_version: version,
             p_engine_name: b"gfx-rs\0".as_ptr() as *const _,
             engine_version: 1,
-            api_version: 0, //TODO
+            api_version: 0, //TODO: VK_API_VERSION_1_1
```
Comment (msiglreith) [inline src/backend/vulkan/src/lib.rs]: heh

Comment (msiglreith) [reply src/backend/vulkan/src/lib.rs]: Actually, 0 is fine

File: src/hal/src/image.rs [hunk 2]
```diff
@@ -254,98 +225,133 @@ pub const CUBE_FACES: [CubeFace; 6] = [
...
-    Cube(Size),
-    /// An array of Cube textures.
-    CubeArray(Size, Layer),
 }
 
 impl Kind {
-    /// Get texture dimensions
-    pub fn dimensions(&self) -> Dimensions {
-        let s0 = AaMode::Single;
+    /// Get the image extent.
```
Comment (msiglreith) [inline src/hal/src/image.rs]: Nit: annotation that these are the extent of a single image layer.

File: src/backend/vulkan/src/lib.rs (hunk 1 above)
Comment (kvark) [reply src/backend/vulkan/src/lib.rs]: it spams a warning to console now (unrelated to the PR): 
> Warning: Unrecognized CreateInstance->pCreateInfo->pApplicationInfo.apiVersion number -- (0x00000000) assuming VK_API_VERSION_1_1.

Comment (msiglreith) [reply src/backend/vulkan/src/lib.rs]: Sounds like a validation layer bug

**Comment by kvark** [comment]
bors r=msiglreith

**Pull request merged**